{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Description\n",
    "#### author Houcmeddine Othman, September 2019, [My Github page](https://github.com/hothman) \n",
    "\n",
    "The codes in this notebook are written to generate the structure of the database accroding to the ERD including the columns for primary keys and foreign keys. The output are 7 tables in csv format. \n",
    "\n",
    "![title](diagram_PM.png)\n",
    "\n",
    "### requirements:\n",
    "* Python 3\n",
    "* Pandas, numpy, pubmed_parser\n",
    "---\n",
    "## Credits\n",
    "### Without the contribution of these people, I would end up with ugly empty output files and big ERROR message. \n",
    "\n",
    "#### Manual Curation (in alphabetical order) \n",
    "Ayoub Ksouri, Chaimae SAMTAL, Chiamaka Jessica Okeke, Fouzia Radouani, Haifa Jmal, Kais Ghedira, Lyndon Zass, Melek Chaouch, Olivier, Reem Sallam, Rym Kefi, Samah Ahmed, Samar kamel Kassem, Yosr Hamdi.\n",
    "\n",
    "#### Data mining and wrangling\n",
    "Jorge da Rocha and Lyndon Zass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Completing P-value to SNP and Star tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 431,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>allele</th>\n",
       "      <th>reference_id</th>\n",
       "      <th>source</th>\n",
       "      <th>id_in_source</th>\n",
       "      <th>P-value</th>\n",
       "      <th>ID_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>rs28371685</td>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>Allele T is associated with decreased dose of ...</td>\n",
       "      <td>T</td>\n",
       "      <td>20072124</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>608431789</td>\n",
       "      <td>&lt; 0.001</td>\n",
       "      <td>SNP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>rs28371686</td>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>Allele G is associated with decreased dose of ...</td>\n",
       "      <td>G</td>\n",
       "      <td>20072124</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>608431793</td>\n",
       "      <td>&lt; 0.001</td>\n",
       "      <td>SNP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>rs7900194</td>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>Allele A is associated with decreased dose of ...</td>\n",
       "      <td>A</td>\n",
       "      <td>20072124</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>608431781</td>\n",
       "      <td>0.023</td>\n",
       "      <td>SNP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>rs9332131</td>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>Genotype DEL is associated with decreased dose...</td>\n",
       "      <td>DEL</td>\n",
       "      <td>20072124</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>608431785</td>\n",
       "      <td>&lt; 0.001</td>\n",
       "      <td>SNP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>rs339097</td>\n",
       "      <td>CALU</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>Allele G is associated with increased dose of ...</td>\n",
       "      <td>G</td>\n",
       "      <td>20200517</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>637879876</td>\n",
       "      <td>0.03</td>\n",
       "      <td>SNP5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>362</td>\n",
       "      <td>rs1902023</td>\n",
       "      <td>UGT2B15</td>\n",
       "      <td>acetaminophen</td>\n",
       "      <td>Genotype AA is associated with increased metab...</td>\n",
       "      <td>AA</td>\n",
       "      <td>28663312</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1448639952</td>\n",
       "      <td>ambiguous</td>\n",
       "      <td>SNP363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>363</td>\n",
       "      <td>CYP3A5</td>\n",
       "      <td>CYP3A5 (PA131)</td>\n",
       "      <td>acetaminophen</td>\n",
       "      <td>Allele C is not associated with metabolism of ...</td>\n",
       "      <td>C</td>\n",
       "      <td>28663312</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1448639975</td>\n",
       "      <td>&gt; 0.05</td>\n",
       "      <td>SNP364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>364</td>\n",
       "      <td>rs8330</td>\n",
       "      <td>UGT1A</td>\n",
       "      <td>acetaminophen</td>\n",
       "      <td>Allele C is not associated with metabolism of ...</td>\n",
       "      <td>C</td>\n",
       "      <td>28663312</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1448639969</td>\n",
       "      <td>&gt; 0.05</td>\n",
       "      <td>SNP365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>365</td>\n",
       "      <td>rs5333</td>\n",
       "      <td>EDNRA</td>\n",
       "      <td>prednisone</td>\n",
       "      <td>Genotypes CC + CT is associated with increased...</td>\n",
       "      <td>CC + CT</td>\n",
       "      <td>30672385</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1450367980</td>\n",
       "      <td>0.016</td>\n",
       "      <td>SNP366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>366</td>\n",
       "      <td>rs2273897</td>\n",
       "      <td>ABCC2</td>\n",
       "      <td>emtricitabine</td>\n",
       "      <td>Genotype TT is associated with increased conce...</td>\n",
       "      <td>TT</td>\n",
       "      <td>30767719</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1450377018</td>\n",
       "      <td>0.028</td>\n",
       "      <td>SNP367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>367 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          rs_id       gene_name      drug_name  \\\n",
       "0    rs28371685          CYP2C9       warfarin   \n",
       "1    rs28371686          CYP2C9       warfarin   \n",
       "2     rs7900194          CYP2C9       warfarin   \n",
       "3     rs9332131          CYP2C9       warfarin   \n",
       "4      rs339097            CALU       warfarin   \n",
       "..          ...             ...            ...   \n",
       "362   rs1902023         UGT2B15  acetaminophen   \n",
       "363      CYP3A5  CYP3A5 (PA131)  acetaminophen   \n",
       "364      rs8330           UGT1A  acetaminophen   \n",
       "365      rs5333           EDNRA     prednisone   \n",
       "366   rs2273897           ABCC2  emtricitabine   \n",
       "\n",
       "                                             phenotype   allele  reference_id  \\\n",
       "0    Allele T is associated with decreased dose of ...        T      20072124   \n",
       "1    Allele G is associated with decreased dose of ...        G      20072124   \n",
       "2    Allele A is associated with decreased dose of ...        A      20072124   \n",
       "3    Genotype DEL is associated with decreased dose...      DEL      20072124   \n",
       "4    Allele G is associated with increased dose of ...        G      20200517   \n",
       "..                                                 ...      ...           ...   \n",
       "362  Genotype AA is associated with increased metab...       AA      28663312   \n",
       "363  Allele C is not associated with metabolism of ...        C      28663312   \n",
       "364  Allele C is not associated with metabolism of ...        C      28663312   \n",
       "365  Genotypes CC + CT is associated with increased...  CC + CT      30672385   \n",
       "366  Genotype TT is associated with increased conce...       TT      30767719   \n",
       "\n",
       "       source  id_in_source    P-value  ID_snp  \n",
       "0    PharmGKB     608431789    < 0.001    SNP1  \n",
       "1    PharmGKB     608431793    < 0.001    SNP2  \n",
       "2    PharmGKB     608431781      0.023    SNP3  \n",
       "3    PharmGKB     608431785    < 0.001    SNP4  \n",
       "4    PharmGKB     637879876       0.03    SNP5  \n",
       "..        ...           ...        ...     ...  \n",
       "362  PharmGKB    1448639952  ambiguous  SNP363  \n",
       "363  PharmGKB    1448639975     > 0.05  SNP364  \n",
       "364  PharmGKB    1448639969     > 0.05  SNP365  \n",
       "365  PharmGKB    1450367980      0.016  SNP366  \n",
       "366  PharmGKB    1450377018      0.028  SNP367  \n",
       "\n",
       "[367 rows x 10 columns]"
      ]
     },
     "execution_count": 431,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd \n",
    "import warnings\n",
    "\n",
    "# read tables \n",
    "snp = pd.read_csv(\"../data/snp_raw.csv\")\n",
    "star = pd.read_csv(\"../data/star_raw.csv\")\n",
    "curation = pd.read_csv(\"../data/clean_curation.csv\")\n",
    "\n",
    "# Some variants were discarded in in the curation table, we delete them also from snp and star\n",
    "ids_in_curation = list( curation.id_in_source )\n",
    "snp = snp[snp.id_in_source.isin(ids_in_curation)]\n",
    "star = star[star.id_in_source.isin(ids_in_curation)]\n",
    "\n",
    "# left joining curation data to snp table\n",
    "snp = pd.merge(snp, curation[[\"id_in_source\", \"P-value\"]] , on=\"id_in_source\")\n",
    "\n",
    "# left joining curation data to star allele table\n",
    "star = pd.merge(star, curation[[\"id_in_source\", \"P-value\"]] , on=\"id_in_source\")\n",
    "\n",
    "# some of the variants are going to be deleted \n",
    "var_to_delete = ['1444665865','1444665876', '1444665886', '1449270328', \n",
    " '1449270311', '1449270340', '1184509852','1184509863', '1448105757']\n",
    "\n",
    "for varID in var_to_delete: \n",
    "    snp = snp[snp.id_in_source != int(varID)]   # panas reads var as integer \n",
    "    star = star[star.id_in_source != int(varID)]\n",
    "    curation = curation[curation.id_in_source != int(varID)]\n",
    "\n",
    "#generating primary keys for snp table \n",
    "len_snp = len(snp)\n",
    "ID_snp = [ \"SNP\"+str(i) for i in range(1, len_snp+1) ]\n",
    "snp[\"ID_snp\"] = ID_snp\n",
    "\n",
    "#generating primary keys for star table \n",
    "len_star = len(star)\n",
    "ID_star = [ \"STAR\"+str(i) for i in range(1, len_star+1) ]\n",
    "star[\"ID_star\"] = ID_star\n",
    "\n",
    "snp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## generating tables ethnicity_country for snp and star \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 432,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_region</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>REGION1</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>REGION2</td>\n",
       "      <td>Mixed Population containing african descendant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>REGION3</td>\n",
       "      <td>North African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>REGION4</td>\n",
       "      <td>Sub-Saharan African</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id_region                                             region\n",
       "0   REGION1                    African American/Afro-Caribbean\n",
       "1   REGION2  Mixed Population containing african descendant...\n",
       "2   REGION3                                      North African\n",
       "3   REGION4                                Sub-Saharan African"
      ]
     },
     "execution_count": 432,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions= list( curation.groupby('region').count().index )\n",
    "len_regions = len(regions)\n",
    "ID_region = [ \"REGION\"+str(i) for i in range(1, len_regions+1) ]\n",
    "region_table = pd.DataFrame({\"id_region\":ID_region, \"region\":regions })\n",
    "\n",
    "region_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## generating table for reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 433,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_id</th>\n",
       "      <th>type</th>\n",
       "      <th>year</th>\n",
       "      <th>title</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ref_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>REF1</td>\n",
       "      <td>24024897</td>\n",
       "      <td>PMID</td>\n",
       "      <td>2013</td>\n",
       "      <td>RRM1 and RRM2 pharmacogenetics: association wi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>REF2</td>\n",
       "      <td>23173844</td>\n",
       "      <td>PMID</td>\n",
       "      <td>2012</td>\n",
       "      <td>PXR and CAR single nucleotide polymorphisms in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>REF3</td>\n",
       "      <td>27045425</td>\n",
       "      <td>PMID</td>\n",
       "      <td>2016</td>\n",
       "      <td>CYP2B6 genotype-based efavirenz dose recommend...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>REF4</td>\n",
       "      <td>15094935</td>\n",
       "      <td>PMID</td>\n",
       "      <td>2004</td>\n",
       "      <td>CYP2C9 genetic polymorphisms and warfarin.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>REF5</td>\n",
       "      <td>23158458</td>\n",
       "      <td>PMID</td>\n",
       "      <td>2013</td>\n",
       "      <td>Multiple regulatory variants modulate expressi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        reference_id  type  year  \\\n",
       "ref_id                             \n",
       "REF1        24024897  PMID  2013   \n",
       "REF2        23173844  PMID  2012   \n",
       "REF3        27045425  PMID  2016   \n",
       "REF4        15094935  PMID  2004   \n",
       "REF5        23158458  PMID  2013   \n",
       "\n",
       "                                                    title  \n",
       "ref_id                                                     \n",
       "REF1    RRM1 and RRM2 pharmacogenetics: association wi...  \n",
       "REF2    PXR and CAR single nucleotide polymorphisms in...  \n",
       "REF3    CYP2B6 genotype-based efavirenz dose recommend...  \n",
       "REF4           CYP2C9 genetic polymorphisms and warfarin.  \n",
       "REF5    Multiple regulatory variants modulate expressi...  "
      ]
     },
     "execution_count": 433,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference = pd.read_csv(\"../data/referance_raw.csv\")\n",
    "len_reference = len(reference)\n",
    "ID_reference = [ \"REF\"+str(i) for i in range(1, len_reference+1) ]\n",
    "reference[\"ref_id\"]=ID_reference\n",
    "reference.rename(columns={'external_id': \"reference_id\"}, inplace=True)\n",
    "reference.set_index(\"ref_id\", inplace=True)\n",
    "\n",
    "reference.to_csv(\"../data/csv4SQL/Study.csv\", index=False)\n",
    "\n",
    "\n",
    "reference.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating pharmacogenes table \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Uniprot_ID</th>\n",
       "      <th>Function</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID_pharmacogenes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>PHARGENE1</td>\n",
       "      <td>7</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>P08183</td>\n",
       "      <td>Translocates drugs and phospholipids across th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PHARGENE2</td>\n",
       "      <td>2</td>\n",
       "      <td>ABCB11</td>\n",
       "      <td>O95342</td>\n",
       "      <td>Catalyzes the secretion of conjugated bile sal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PHARGENE3</td>\n",
       "      <td>10</td>\n",
       "      <td>ABCC2</td>\n",
       "      <td>Q92887</td>\n",
       "      <td>Mediates hepatobiliary excretion of numerous o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PHARGENE4</td>\n",
       "      <td>4</td>\n",
       "      <td>ABCG2</td>\n",
       "      <td>Q9UNQ0</td>\n",
       "      <td>High-capacity urate exporter functioning in bo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PHARGENE5</td>\n",
       "      <td>17</td>\n",
       "      <td>ACE</td>\n",
       "      <td>P12821</td>\n",
       "      <td>Converts angiotensin I to angiotensin II by re...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Chromosome gene_name Uniprot_ID  \\\n",
       "ID_pharmacogenes                                   \n",
       "PHARGENE1                 7     ABCB1     P08183   \n",
       "PHARGENE2                 2    ABCB11     O95342   \n",
       "PHARGENE3                10     ABCC2     Q92887   \n",
       "PHARGENE4                 4     ABCG2     Q9UNQ0   \n",
       "PHARGENE5                17       ACE     P12821   \n",
       "\n",
       "                                                           Function  \n",
       "ID_pharmacogenes                                                     \n",
       "PHARGENE1         Translocates drugs and phospholipids across th...  \n",
       "PHARGENE2         Catalyzes the secretion of conjugated bile sal...  \n",
       "PHARGENE3         Mediates hepatobiliary excretion of numerous o...  \n",
       "PHARGENE4         High-capacity urate exporter functioning in bo...  \n",
       "PHARGENE5         Converts angiotensin I to angiotensin II by re...  "
      ]
     },
     "execution_count": 434,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pharmacogenes = pd.read_csv(\"../data/genes_prot_chromosome_function.csv\")\n",
    "len_pharmacogenes = len(pharmacogenes)\n",
    "ID_pharmacogenes = [ \"PHARGENE\"+str(i) for i in range(1, len_pharmacogenes+1) ]\n",
    "pharmacogenes[\"ID_pharmacogenes\"] = ID_pharmacogenes \n",
    "pharmacogenes.set_index(\"ID_pharmacogenes\", inplace=True)\n",
    "\n",
    "pharmacogenes.to_csv(\"../data/csv4SQL/pharmacogenes.csv\", index=False)\n",
    "\n",
    "pharmacogenes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating drug table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_name</th>\n",
       "      <th>ID Drug bank</th>\n",
       "      <th>state</th>\n",
       "      <th>Indication</th>\n",
       "      <th>IUPAC_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID_drugs</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>DRUG1</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>DB00682</td>\n",
       "      <td>Approved</td>\n",
       "      <td>Prophylaxis and treatment of venous thromboemb...</td>\n",
       "      <td>4-hydroxy-3-(3-oxo-1-phenylbutyl)-2H-chromen-2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DRUG2</td>\n",
       "      <td>haloperidol</td>\n",
       "      <td>DB00502</td>\n",
       "      <td>Approved</td>\n",
       "      <td>Haloperidol is indicated for a number of condi...</td>\n",
       "      <td>4-[4-(4-chlorophenyl)-4-hydroxypiperidin-1-yl]...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DRUG3</td>\n",
       "      <td>clozapine</td>\n",
       "      <td>DB00363</td>\n",
       "      <td>Approved</td>\n",
       "      <td>For use in patients with treatment-resistant s...</td>\n",
       "      <td>8-Chloro-11-(4-methylpiperazin-1-yl)-5H-dibenz...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DRUG4</td>\n",
       "      <td>mephenytoin</td>\n",
       "      <td>DB00532</td>\n",
       "      <td>Approved, Investigational, Withdrawn</td>\n",
       "      <td>For the treatment of refractory partial epilepsy</td>\n",
       "      <td>5-ethyl-3-methyl-5-phenylimidazolidine-2,4-dione</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DRUG5</td>\n",
       "      <td>efavirenz</td>\n",
       "      <td>DB00625</td>\n",
       "      <td>Approved, Investigational</td>\n",
       "      <td>(4S)-6-chloro-4-(2-cyclopropylethynyl)-4-(trif...</td>\n",
       "      <td>(4S)-6-Chloro-4-(cyclopropylethynyl)-1,4-dihyd...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            drug_name ID Drug bank                                 state  \\\n",
       "ID_drugs                                                                   \n",
       "DRUG1        warfarin      DB00682                              Approved   \n",
       "DRUG2     haloperidol      DB00502                              Approved   \n",
       "DRUG3       clozapine      DB00363                              Approved   \n",
       "DRUG4     mephenytoin      DB00532  Approved, Investigational, Withdrawn   \n",
       "DRUG5       efavirenz      DB00625             Approved, Investigational   \n",
       "\n",
       "                                                 Indication  \\\n",
       "ID_drugs                                                      \n",
       "DRUG1     Prophylaxis and treatment of venous thromboemb...   \n",
       "DRUG2     Haloperidol is indicated for a number of condi...   \n",
       "DRUG3     For use in patients with treatment-resistant s...   \n",
       "DRUG4      For the treatment of refractory partial epilepsy   \n",
       "DRUG5     (4S)-6-chloro-4-(2-cyclopropylethynyl)-4-(trif...   \n",
       "\n",
       "                                                 IUPAC_name  \n",
       "ID_drugs                                                     \n",
       "DRUG1     4-hydroxy-3-(3-oxo-1-phenylbutyl)-2H-chromen-2...  \n",
       "DRUG2     4-[4-(4-chlorophenyl)-4-hydroxypiperidin-1-yl]...  \n",
       "DRUG3     8-Chloro-11-(4-methylpiperazin-1-yl)-5H-dibenz...  \n",
       "DRUG4      5-ethyl-3-methyl-5-phenylimidazolidine-2,4-dione  \n",
       "DRUG5     (4S)-6-Chloro-4-(cyclopropylethynyl)-1,4-dihyd...  "
      ]
     },
     "execution_count": 442,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugs = pd.read_csv(\"../data/drugs.csv\")\n",
    "len_drugs = len(drugs)\n",
    "ID_drugs = [ \"DRUG\"+str(i) for i in range(1, len_drugs+1) ]\n",
    "drugs[\"ID_drugs\"] = ID_drugs \n",
    "drugs.set_index(\"ID_drugs\", inplace=True)\n",
    "drugs.rename(columns={'Drug name': \"drug_name\"}, inplace=True)\n",
    "\n",
    "drugs.to_csv(\"../data/csv4SQL/drug.csv\", index=False)\n",
    "\n",
    "drugs.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Arranging secondary keys for SQL implementation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_col_from_index(df_on_left, df_On_right, col_to_merge, tag ):\n",
    "    mydf = df_On_right.copy()\n",
    "    index_col = mydf.index\n",
    "    mydf[\"index_col\"] = index_col    \n",
    "    mydf=mydf.filter(items=['index_col', col_to_merge ] )  \n",
    "    merged= pd.merge(df_on_left, mydf, on=col_to_merge).drop(columns=[col_to_merge])\n",
    "    return merged.rename(columns={'index_col': tag} )\n",
    "\n",
    "\n",
    "# add gene id\n",
    "snp = create_col_from_index(snp, pharmacogenes , \"gene_name\", \"gene_id\" )\n",
    "\n",
    "# add drug id, corrected later aftergenerating the whole able\n",
    "snp = create_col_from_index(snp, drugs , \"drug_name\", \"drug_id\" )\n",
    "\n",
    "# add reference id \n",
    "snp = create_col_from_index(snp, reference , \"reference_id\", \"reference_id\" ) \n",
    "\n",
    "\n",
    "# arranging columns \n",
    "snp.rename(columns={'ID_snp': \"id\"}, inplace=True )\n",
    "\n",
    "snp= snp.filter(items=['id', \"rs_id\", \"drug_id\", \n",
    "                  \"allele\", \"phenotype\", \n",
    "                  \"reference_id\", \"P-value\", \n",
    "                 \"source\", \"id_in_source\"])\n",
    "\n",
    "snp.to_csv(\"../data/csv4SQL/snp_allele.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Star allele table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 437,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add gene id\n",
    "star= create_col_from_index(star, pharmacogenes , \"gene_name\", \"gene_id\" )\n",
    "# add drug id, corrected later aftergenerating the whole able\n",
    "star = create_col_from_index(star, drugs , \"drug_name\", \"drug_id\" )\n",
    "# add reference id \n",
    "star = create_col_from_index(star, reference , \"reference_id\", \"reference_id\" ) \n",
    "# arranging columns \n",
    "star.rename(columns={'ID_star': \"id\"}, inplace=True )\n",
    "\n",
    "star= star.filter(items=['id', \"star_annotation\", \"drug_id\", \n",
    "                  \"allele\", \"phenotype\", \n",
    "                  \"reference_id\", \"P-value\", \n",
    "                 \"source\", \"id_in_source\"])\n",
    "\n",
    "\n",
    "star.to_csv(\"../data/csv4SQL/star_allele.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### snp by country table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>region</th>\n",
       "      <th>Country_of_Participants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>SNPREGION1</td>\n",
       "      <td>SNP1</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>SNPREGION2</td>\n",
       "      <td>SNP1</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>SNPREGION3</td>\n",
       "      <td>SNP2</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>SNPREGION4</td>\n",
       "      <td>SNP2</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>SNPREGION5</td>\n",
       "      <td>SNP3</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           id snp_id                           region Country_of_Participants\n",
       "0  SNPREGION1   SNP1  African American/Afro-Caribbean                     USA\n",
       "1  SNPREGION2   SNP1  African American/Afro-Caribbean                     USA\n",
       "2  SNPREGION3   SNP2  African American/Afro-Caribbean                     USA\n",
       "3  SNPREGION4   SNP2  African American/Afro-Caribbean                     USA\n",
       "4  SNPREGION5   SNP3  African American/Afro-Caribbean                     USA"
      ]
     },
     "execution_count": 438,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_countries = pd.merge(snp, curation, on=\"id_in_source\")\n",
    "merged_countries_snp = merged_countries.filter(items=[\"id\", \"region\", \"Country_of_Participants\"]) \n",
    "merged_countries_snp.rename(columns={'id': \"snp_id\"}, inplace=True )\n",
    "len_merged_countries_snp = len(merged_countries_snp)\n",
    "ID_countries_snp = [ \"SNPREGION\"+str(i) for i in range(1, len_merged_countries_snp+1) ]\n",
    "merged_countries_snp[\"id\"]=ID_countries_snp\n",
    "merged_countries_snp = merged_countries_snp.filter(items=[\"id\", \"snp_id\", \"region\", \"Country_of_Participants\"])\n",
    "\n",
    "# output table \n",
    "merged_countries_snp.to_csv(\"../data/csv4SQL/snp_country.csv\", index=False)\n",
    "\n",
    "merged_countries_snp.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### star by country "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>star_id</th>\n",
       "      <th>region</th>\n",
       "      <th>Country_of_Participants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>STARREGION1</td>\n",
       "      <td>STAR1</td>\n",
       "      <td>Sub-Saharan African</td>\n",
       "      <td>Ethiopia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>STARREGION2</td>\n",
       "      <td>STAR2</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>STARREGION3</td>\n",
       "      <td>STAR3</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>STARREGION4</td>\n",
       "      <td>STAR196</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>STARREGION5</td>\n",
       "      <td>STAR197</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            id  star_id                           region  \\\n",
       "0  STARREGION1    STAR1              Sub-Saharan African   \n",
       "1  STARREGION2    STAR2  African American/Afro-Caribbean   \n",
       "2  STARREGION3    STAR3  African American/Afro-Caribbean   \n",
       "3  STARREGION4  STAR196  African American/Afro-Caribbean   \n",
       "4  STARREGION5  STAR197  African American/Afro-Caribbean   \n",
       "\n",
       "  Country_of_Participants  \n",
       "0                Ethiopia  \n",
       "1                     USA  \n",
       "2                     USA  \n",
       "3                     USA  \n",
       "4                     USA  "
      ]
     },
     "execution_count": 439,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_countries = pd.merge(star, curation, on=\"id_in_source\")\n",
    "merged_countries_star = merged_countries.filter(items=[\"id\", \"region\", \"Country_of_Participants\"]) \n",
    "merged_countries_star.rename(columns={'id': \"star_id\"}, inplace=True )\n",
    "len_merged_countries_star = len(merged_countries_star)\n",
    "ID_countries_star = [ \"STARREGION\"+str(i) for i in range(1, len_merged_countries_star+1) ]\n",
    "merged_countries_star[\"id\"]=ID_countries_star\n",
    "merged_countries_star = merged_countries_star.filter(items=[\"id\", \"star_id\", \"region\", \"Country_of_Participants\"])\n",
    "\n",
    "# output the table\n",
    "merged_countries_star.to_csv(\"../data/csv4SQL/star_allele_country.csv\", index=False)\n",
    "\n",
    "merged_countries_star.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
